vnext/ReactUWP/Views/KeyboardEventHandler.cpp

```diff
@@ -0,0 +1,317 @@
+}
+
+void KeyboardEventHandler::unhook()
+{
+  m_KeyDownRevoker.revoke();
+  m_KeyUpRevoker.revoke();
+}
+
+PreviewKeyboardEventHandlerOnRoot::PreviewKeyboardEventHandlerOnRoot(const std::weak_ptr<IReactInstance>& reactInstance)
+  :PreviewKeyboardEventHandler(
+    std::bind(&PreviewKeyboardEventHandlerOnRoot::OnPreKeyDown, this, _1, _2),
+    std::bind(&PreviewKeyboardEventHandlerOnRoot::OnPreKeyUp, this, _1, _2)),
+  m_wkReactInstance(reactInstance)
+{
+}
+
+void PreviewKeyboardEventHandlerOnRoot::OnPreKeyDown(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const&args)
+{
+  DispatchEventToJs("topKeyDown", args);
+}
+
+void PreviewKeyboardEventHandlerOnRoot::OnPreKeyUp(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const&args)
+{
+  DispatchEventToJs("topKeyUp", args);
+}
+
+inline bool IsModifiedKeyPressed(winrt::CoreWindow const& coreWindow, winrt::VirtualKey vk)
+{
+  return (coreWindow.GetKeyState(vk) & winrt::CoreVirtualKeyStates::Down) == winrt::CoreVirtualKeyStates::Down;
+}
+
+inline bool IsModifiedKeyLocked(winrt::CoreWindow const& coreWindow, winrt::VirtualKey vk)
+{
+  return (coreWindow.GetKeyState(vk) & winrt::CoreVirtualKeyStates::Locked) == winrt::CoreVirtualKeyStates::Locked;
+}
+
+template<typename T> void UpdateModifiedKeyStatusTo(T& ev)
+{
+  auto const& coreWindow = winrt::CoreWindow::GetForCurrentThread();
+  ev.altKey = IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::Menu);
+  ev.shiftKey = IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::Shift);
+  ev.metaKey = IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::LeftWindows)
+    || IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::RightWindows);
+  ev.ctrlKey = IsModifiedKeyPressed(coreWindow, winrt::VirtualKey::Control);
+  ev.capLocked = IsModifiedKeyLocked(coreWindow, winrt::VirtualKey::CapitalLock);
+};
+
+void PreviewKeyboardEventHandlerOnRoot::DispatchEventToJs(string eventName, winrt::KeyRoutedEventArgs args)
+{
+  if (auto instance = m_wkReactInstance.lock())
+  {
+    if (auto source = args.OriginalSource().try_as<winrt::FrameworkElement>())
+    {
+      auto reactId = getReactId(instance.operator->(), source);
+      if (reactId.isValid)
+      {
+        ReactKeyboardEvent ev;
+        ev.target = reactId.tag;
+        UpdateModifiedKeyStatusTo(ev);
+        ev.key = KeyboardHelper::FromVirutalKey(args.Key(), ev.shiftKey, ev.capLocked);
+        instance->DispatchEvent(ev.target, eventName, ToEventData(ev));
+      }
+    }
+  }
+}
+
+KeyboardEvent KeyboardHelper::CreateKeyboardEvent(HandledEventPhase phase, winrt::KeyRoutedEventArgs const& args)
+{
+  KeyboardEvent ev;
+  ev.handledEventPhase = phase;
+  UpdateModifiedKeyStatusTo(ev);
+  ev.key = KeyboardHelper::FromVirutalKey(args.Key(), ev.shiftKey, ev.capLocked);
+
+  return ev;
+}
+
+// Should align to https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values
+static const std::map<winrt::VirtualKey, string> g_virtualKeyToString
```

Comment:
>std::map [](start = 13, length = 8)

Consider using std::vector to store this collection - with collections this small, std::vector typically outperforms std::map due to better cache locality.

Reply:
Keep existing implementation.
map will be more readable here. otherwise I need to create a new structure to keep item if I go with vector.

Reply:
This should be as simple as declaring it as vector<pair<VirtualKey, string>>

---
In reply to: [295538482](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295538482) [](ancestors = 295538482)

Reply:
I didn't see a positive resolution to this.  I'd really prefer you use vector here.

---
In reply to: [295541688](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295541688) [](ancestors = 295541688,295538482)

Reply:
Although I would really like to see this improvement made, we have limited time before 6/30, feel free to track this with a bug.

---
In reply to: [295594796](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295594796) [](ancestors = 295594796,295541688,295538482)

Reply:
Done although I disagree with you

---
In reply to: [296416592](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r296416592) [](ancestors = 296416592,295594796,295541688,295538482)